src/solution.rs

```diff
@@ -1,18 +1,71 @@
-mod suite_run;
 mod test_run;
 
+use std::io::Write;
 use std::process::Command;
 use std::time::Duration;
 
-use suite_run::SuiteRun;
 pub use test_run::{TestResult, TestRun};
+use wait_timeout::ChildExt;
 
 use crate::clash::TestCase;
 
-pub fn run(
-    testcases: Vec<&TestCase>,
-    run_command: Command,
-    timeout: Duration,
-) -> impl IntoIterator<Item = TestRun> {
-    SuiteRun::new(testcases, run_command, timeout)
+pub fn lazy_run<'a>(
+    testcases: impl IntoIterator<Item = &'a TestCase>,
+    run_command: &'a mut Command,
+    timeout: &'a Duration,
+) -> impl IntoIterator<Item = TestRun<'a>> {
+    testcases.into_iter().map(|test| run_testcase(test, run_command, timeout))
+}
+
+fn run_testcase<'a>(test: &'a TestCase, run_command: &mut Command, timeout: &Duration) -> TestRun<'a> {
+    let mut run = match run_command
+        .stdin(std::process::Stdio::piped())
+        .stdout(std::process::Stdio::piped())
+        .stderr(std::process::Stdio::piped())
+        .spawn()
+    {
+        Ok(run) => run,
+        Err(error) => return TestRun::new(test, unable_to_run(error, run_command)),
+    };
+
+    run.stdin
+        .as_mut()
+        .unwrap()
+        .write(test.test_in.as_bytes())
+        .expect("Fatal error: could not write to stdin.");
+
+    TestRun::new(test, get_result(run, &test.test_out, timeout))
+}
+
+fn unable_to_run(error: std::io::Error, cmd: &mut Command) -> TestResult {
+    TestResult::UnableToRun {
+        error_msg: format!("{}: {}", cmd.get_program().to_str().unwrap_or("Unable to run command"), error),
+    }
+}
+
+fn get_result(mut run: std::process::Child, expected: &str, timeout: &Duration) -> TestResult {
+    let timed_out = run.wait_timeout(*timeout).expect("Could not wait for program execution.").is_none();
+
+    if timed_out {
+        run.kill().expect("Failed to kill test run process");
```

Comment:
I know the original doesn't follow the convention either but since the code was originally written I've learned about https://doc.rust-lang.org/std/result/enum.Result.html#recommended-message-style

Reply:
Interesting, I didn't know about that

Reply:
I always thought of expect's message having the purpose of displaying useful error messages but from that it looks like it's intended to document why you are unwrapping the value.